pkg/envoy/streamctrl.go

```diff
@@ -38,6 +46,58 @@ type StreamControl struct {
 	sentVersion             uint64
 	currentVersion          uint64
 	currentVersionNackCount uint
+	completions             []versionCompletion
+}
+
+func (ctrl *StreamControl) addCompletion(completions policy.CompletionContainer, msg string) {
+	if completions == nil {
+		return
+	}
+
+	comp, timeout := completions.AddCompletion()
+
+	// Note that we do not start a timer for the timeout, but rely on NACKs to be followed by
+	// retries that allows for checking for timeouts at some times in future. This should
+	// also work accross Envoy restarts.
+	if comp != nil {
+		log.Debug("Envoy: AddCompletion: ", msg)
+		ctrl.completions = append(ctrl.completions,
+			versionCompletion{ctrl.currentVersion, msg, time.Now().Add(timeout), comp})
+	}
+}
+
+// Called with ctrl.cond.L.Lock() held
+func (ctrl *StreamControl) handleCompletions(version uint64, success bool) {
+	now := time.Now()
+
+	for idx := 0; idx < len(ctrl.completions); {
+		result := success
+
+		if version >= ctrl.completions[idx].version {
+			// Ack or Nack for this version OR later received
+		} else if now.After(ctrl.completions[idx].expiration) {
+			// Timed out, return failure.
+			result = false
+		} else {
+			// Proceed to the next entry
+			idx++
+			continue
+		}
+
+		comp := ctrl.completions[idx]
+
+		res := "NACK"
+		if result {
+			res = "ACK"
+		}
+		log.Debug("Envoy: ", ctrl.name, " ", comp.msg, " ", res, ", time left: ", comp.expiration.Sub(now))
+
+		// Remove the entry
+		ctrl.completions[idx] = ctrl.completions[len(ctrl.completions)-1]
```

Comment:
This is a weird construction IMO, you are effectively shrinking `ctrl.completions` slice until you processed all entries.

If `ctrl.completions` can be changed after going through this loop I would suggest to rewrite this whole loop as follows, to avoid somewhat convoluted slice filtering:
```
	tmpCompletions := make([]versionCompletion, 0, len(ctrl.completions))
	for _, comp := range ctrl.completions {

		result := success

		if version >= comp.version {
			// Ack or Nack for this version OR later received
		} else if now.After(comp.expiration) {
			// Timed out, return failure.
			result = false
		} else {
			// retain the entry
			append(tmpCompletions, comp)
		}

		res := "NACK"
		if result {
			res = "ACK"
		}
		log.Debug("Envoy: ", ctrl.name, " ", comp.msg, " ", res, ", time left: ", comp.expiration.Sub(now))

		comp.completion.Completed(result)
	}
	ctrl.completions = tmpCompletions
```
Hope this makes sense.

Reply:
Makes sense, even though I don't like the fact that range loop copies the object into a temporary (here `comp`). However, in most cases we do this after a successful Envoy transaction, so it makes no difference in practice.

I also did not include the preallocation of the temporary array, as with that the array never shrinks. While we are not dealing with large arrays, IMO it is a bad habit. Again, in most cases I expect we don't need to retain entries, so this does not make a difference either.

I also had to keep the `continue` to not call completion when an entry is retained.

I do like that this is more readable now!